File: next/kernel/protocol.py
"""Wire-level validation and channel matching for the Viewer bus."""

from __future__ import annotations

import re
import uuid
from collections.abc import Mapping
from typing import Any, TypeGuard

PROTOCOL_VERSION = 1
MAX_DEPTH = 8
DEFAULT_FRAME_SIZE = 1024 * 1024
DEFAULT_OUTBOUND_QUEUE = 1000

FRAME_TYPES = frozenset({"hello", "publish", "set", "subscribe", "unsubscribe"})
_NORMAL_FIELD = re.compile(r"[a-z0-9][a-z0-9_-]*\Z")
_RESERVED_FIELD = re.compile(r"_[a-z0-9_-]*\Z")


class ProtocolError(ValueError):
    """A non-fatal wire protocol validation error."""

    def __init__(self, code: str, message: str, detail: Any | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.detail = detail


class HelloValidationError(ValueError):
    """A fatal hello schema error."""


def _is_object(value: object) -> TypeGuard[dict[str, Any]]:
    return isinstance(value, dict) and all(isinstance(key, str) for key in value)


def _valid_field(field: str, *, first: bool) -> bool:
    if _NORMAL_FIELD.fullmatch(field):
        return True
    if field == "_":
        return True
    return first and _RESERVED_FIELD.fullmatch(field) is not None


def validate_channel(channel: object) -> str:
    """Validate a concrete channel and return it.

    Concrete messages have the protocol's fixed three semantic fields. Reserved
    fields are ``_`` placeholders, or an underscore-prefixed first field such as
    ``_inbox`` and ``_conn``.
    """

    if not isinstance(channel, str):
        raise ProtocolError("invalid_channel", "channel must be a string")
    fields = channel.split(":")
    if len(fields) < 3:
        raise ProtocolError("invalid_channel", "channel must contain at least three fields")
    if any(not _valid_field(field, first=index == 0) for index, field in enumerate(fields)):
        raise ProtocolError("invalid_channel", f"invalid channel: {channel}")
    return channel


def validate_pattern(pattern: object) -> str:
    """Validate a subscription pattern and return it."""

    if not isinstance(pattern, str):
        raise ProtocolError("invalid_pattern", "pattern must be a string")
    if pattern == ">":
        return pattern
    fields = pattern.split(":")
    if not fields or any(not field for field in fields):
        raise ProtocolError("invalid_pattern", f"invalid pattern: {pattern}")
    for index, field in enumerate(fields):
        if field == "*":
            continue
        if ">" in field or "*" in field or not _valid_field(field, first=index == 0):
            raise ProtocolError("invalid_pattern", f"invalid pattern: {pattern}")
    return pattern


def channel_matches(pattern: str, channel: str) -> bool:
    """Implement protocol specification section 4.2 exactly."""

    validate_pattern(pattern)
    validate_channel(channel)
    if pattern == ">":
        return True
    pattern_fields = pattern.split(":")
    channel_fields = channel.split(":")
    if len(pattern_fields) > len(channel_fields):
        return False
    return all(
        pattern_field == "*" or pattern_field == channel_fields[index]
        for index, pattern_field in enumerate(pattern_fields)
    )


def validate_hello(frame: object) -> dict[str, Any]:
    """Strongly validate the first-frame hello schema.

    Version compatibility is deliberately checked by the server so it can use
    the distinct 4003 close code.
    """

    if not _is_object(frame):
        raise HelloValidationError("hello must be a JSON object")
    required = {"type", "protocol_version", "conn", "manifest", "managed"}
    missing = sorted(required - frame.keys())
    if missing:
        raise HelloValidationError(f"hello missing required fields: {', '.join(missing)}")
    if frame["type"] != "hello":
        raise HelloValidationError("frame type must be hello")
    if isinstance(frame["protocol_version"], bool) or not isinstance(frame["protocol_version"], int):
        raise HelloValidationError("protocol_version must be an integer")
    conn = frame["conn"]
    if not isinstance(conn, str):
        raise HelloValidationError("conn must be a UUIDv4 string")
    try:
        parsed_conn = uuid.UUID(conn)
    except (ValueError, AttributeError) as exc:
        raise HelloValidationError("conn must be a UUIDv4 string") from exc
    if parsed_conn.version != 4:
        raise HelloValidationError("conn must be a UUIDv4 string")
    manifest = frame["manifest"]
    if not _is_object(manifest):
        raise HelloValidationError("manifest must be an object")
    manifest_required = {"id", "version", "slots", "emits"}
    manifest_missing = sorted(manifest_required - manifest.keys())
    if manifest_missing:
        raise HelloValidationError(
            f"manifest missing required fields: {', '.join(manifest_missing)}"
        )
    plugin_id = manifest["id"]
    if not isinstance(plugin_id, str) or _NORMAL_FIELD.fullmatch(plugin_id) is None:
        raise HelloValidationError("manifest.id must be a valid non-reserved channel field")
    if not isinstance(manifest["version"], str) or not manifest["version"]:
        raise HelloValidationError("manifest.version must be a non-empty string")
    if not _is_object(manifest["slots"]) or not _is_object(manifest["emits"]):
        raise HelloValidationError("manifest.slots and manifest.emits must be objects")
    if not isinstance(frame["managed"], bool):
        raise HelloValidationError("managed must be a boolean")
    if "instance_id" in frame:
        instance_id = frame["instance_id"]
        if not isinstance(instance_id, str) or not _valid_field(instance_id, first=False):
            raise HelloValidationError("instance_id must be a valid channel field")
    return frame


def validate_post_hello_frame(frame: object) -> dict[str, Any]:
    """Validate protocol metadata while deliberately leaving ``value`` free-form."""

    if not _is_object(frame):
        raise ProtocolError("malformed_frame", "frame must be a JSON object")
    frame_type = frame.get("type")
    if not isinstance(frame_type, str):
        raise ProtocolError("malformed_frame", "frame.type must be a string")
    if frame_type not in FRAME_TYPES:
        raise ProtocolError("unknown_type", f"unknown frame type: {frame_type}")
    if frame_type == "hello":
        raise ProtocolError("unexpected_hello", "hello may only be sent as the first frame")
    if frame_type in {"subscribe", "unsubscribe"}:
        validate_pattern(frame.get("pattern"))
        return frame
    validate_channel(frame.get("channel"))
    if "value" not in frame:
        raise ProtocolError("malformed_frame", f"{frame_type} frame missing value")
    trace_id = frame.get("trace_id")
    if trace_id is not None and (not isinstance(trace_id, str) or not trace_id):
        raise ProtocolError("malformed_frame", "trace_id must be a non-empty string")
    depth = frame.get("depth", 0)
    if isinstance(depth, bool) or not isinstance(depth, int) or depth < 0:
        raise ProtocolError("malformed_frame", "depth must be a non-negative integer")
    return frame


def frame_limit_for(
    frame: Mapping[str, Any] | None,
    overrides: Mapping[str, int],
    default: int = DEFAULT_FRAME_SIZE,
) -> int:
    """Return the largest configured limit matching a frame's channel prefix."""

    if frame is None or not isinstance(frame.get("channel"), str):
        return default
    channel = frame["channel"]
    limits = [limit for prefix, limit in overrides.items() if channel.startswith(prefix)]
    return max([default, *limits])

File: next-go/internal/protocol/protocol.go
package protocol

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strings"
)

const (
	Version           = 1
	MaxDepth          = 8
	DefaultFrameSize  = 1024 * 1024
	DefaultQueueSize  = 1000
	KernelPluginID    = "kernel"
	DefaultInstanceID = "_"
)

var (
	normalField   = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]*$`)
	reservedField = regexp.MustCompile(`^_[a-z0-9_-]*$`)
	uuidV4        = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89aAbB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$`)
)

type Origin struct {
	Plugin   string `json:"plugin"`
	Instance string `json:"instance"`
}

type Manifest struct {
	ID      string          `json:"id"`
	Version string          `json:"version"`
	Slots   map[string]any  `json:"slots"`
	Emits   map[string]any  `json:"emits"`
	Raw     json.RawMessage `json:"-"`
}

func (m Manifest) MarshalJSON() ([]byte, error) {
	if len(m.Raw) != 0 {
		return m.Raw, nil
	}
	type wireManifest Manifest
	return json.Marshal(wireManifest(m))
}

type Hello struct {
	Type            string   `json:"type"`
	ProtocolVersion int      `json:"protocol_version"`
	Conn            string   `json:"conn"`
	Manifest        Manifest `json:"manifest"`
	Managed         bool     `json:"managed"`
	InstanceID      *string  `json:"instance_id,omitempty"`
}

type PublishFrame struct {
	Type    string          `json:"type"`
	Channel string          `json:"channel"`
	Value   json.RawMessage `json:"value"`
	TraceID string          `json:"trace_id,omitempty"`
	Depth   int             `json:"depth,omitempty"`
	TS      int64           `json:"ts,omitempty"`
	Origin  *Origin         `json:"origin,omitempty"`
}

type SetFrame struct {
	Type    string          `json:"type"`
	Channel string          `json:"channel"`
	Value   json.RawMessage `json:"value"`
	TraceID string          `json:"trace_id,omitempty"`
	Depth   int             `json:"depth,omitempty"`
	TS      int64           `json:"ts,omitempty"`
	Origin  *Origin         `json:"origin,omitempty"`
}

type SubscribeFrame struct {
	Type    string `json:"type"`
	Pattern string `json:"pattern"`
}

type UnsubscribeFrame struct {
	Type    string `json:"type"`
	Pattern string `json:"pattern"`
}

type ClientFrame struct {
	Type    string
	Channel string
	Pattern string
	Value   json.RawMessage
	TraceID string
	Depth   int
}

type Delivery struct {
	Type    string          `json:"type"`
	Channel string          `json:"channel"`
	Value   json.RawMessage `json:"value"`
	TraceID string          `json:"trace_id,omitempty"`
	Depth   int             `json:"depth"`
	TS      int64           `json:"ts"`
	Origin  Origin          `json:"origin"`
}

type Error struct {
	Code    string
	Message string
	Detail  any
}

func (e *Error) Error() string { return e.Message }

func validField(field string, first bool) bool {
	return normalField.MatchString(field) || field == "_" || first && reservedField.MatchString(field)
}

func ValidateChannel(channel string) error {
	fields := strings.Split(channel, ":")
	if len(fields) < 3 {
		return &Error{Code: "invalid_channel", Message: "channel must contain at least three fields"}
	}
	for i, field := range fields {
		if !validField(field, i == 0) {
			return &Error{Code: "invalid_channel", Message: fmt.Sprintf("invalid channel: %s", channel)}
		}
	}
	return nil
}

func ValidatePattern(pattern string) error {
	if pattern == ">" {
		return nil
	}
	fields := strings.Split(pattern, ":")
	for i, field := range fields {
		if field == "" {
			return &Error{Code: "invalid_pattern", Message: fmt.Sprintf("invalid pattern: %s", pattern)}
		}
		if field == "*" {
			continue
		}
		if strings.ContainsAny(field, "*>") || !validField(field, i == 0) {
			return &Error{Code: "invalid_pattern", Message: fmt.Sprintf("invalid pattern: %s", pattern)}
		}
	}
	return nil
}

func ChannelMatches(pattern, channel string) bool {
	if pattern == ">" {
		return true
	}
	pf := strings.Split(pattern, ":")
	cf := strings.Split(channel, ":")
	if len(pf) > len(cf) {
		return false
	}
	for i := range pf {
		if pf[i] != "*" && pf[i] != cf[i] {
			return false
		}
	}
	return true
}

func ParseHello(data []byte) (Hello, error) {
	object, err := rawObject(data)
	if err != nil {
		return Hello{}, errors.New("hello must be a JSON object")
	}
	required := []string{"type", "protocol_version", "conn", "manifest", "managed"}
	for _, field := range required {
		if _, ok := object[field]; !ok {
			return Hello{}, fmt.Errorf("hello missing required field: %s", field)
		}
	}
	var hello Hello
	if err := json.Unmarshal(data, &hello); err != nil {
		return Hello{}, errors.New("hello fields have invalid types")
	}
	if hello.Type != "hello" {
		return Hello{}, errors.New("frame type must be hello")
	}
	if !uuidV4.MatchString(hello.Conn) {
		return Hello{}, errors.New("conn must be a UUIDv4 string")
	}
	manifestRaw, ok := object["manifest"]
	if !ok {
		return Hello{}, errors.New("manifest must be an object")
	}
	manifestObject, err := rawObject(manifestRaw)
	if err != nil {
		return Hello{}, errors.New("manifest must be an object")
	}
	for _, field := range []string{"id", "version", "slots", "emits"} {
		if _, ok := manifestObject[field]; !ok {
			return Hello{}, fmt.Errorf("manifest missing required field: %s", field)
		}
	}
	if !normalField.MatchString(hello.Manifest.ID) {
		return Hello{}, errors.New("manifest.id must be a valid non-reserved channel field")
	}
	if hello.Manifest.Version == "" {
		return Hello{}, errors.New("manifest.version must be a non-empty string")
	}
	if hello.Manifest.Slots == nil || hello.Manifest.Emits == nil {
		return Hello{}, errors.New("manifest.slots and manifest.emits must be objects")
	}
	hello.Manifest.Raw = append(json.RawMessage(nil), manifestRaw...)
	if hello.InstanceID != nil && !validField(*hello.InstanceID, false) {
		return Hello{}, errors.New("instance_id must be a valid channel field")
	}
	return hello, nil
}

func ParseClientFrame(data []byte) (ClientFrame, error) {
	object, err := rawObject(data)
	if err != nil {
		return ClientFrame{}, &Error{Code: "malformed_frame", Message: "frame must be a JSON object"}
	}
	typeRaw, ok := object["type"]
	if !ok {
		return ClientFrame{}, &Error{Code: "malformed_frame", Message: "frame.type must be a string"}
	}
	var frameType string
	if err := json.Unmarshal(typeRaw, &frameType); err != nil {
		return ClientFrame{}, &Error{Code: "malformed_frame", Message: "frame.type must be a string"}
	}
	if frameType == "hello" {
		return ClientFrame{}, &Error{Code: "unexpected_hello", Message: "hello may only be sent as the first frame"}
	}
	if frameType != "publish" && frameType != "set" && frameType != "subscribe" && frameType != "unsubscribe" {
		return ClientFrame{}, &Error{Code: "unknown_type", Message: fmt.Sprintf("unknown frame type: %s", frameType)}
	}
	frame := ClientFrame{Type: frameType}
	if frameType == "subscribe" || frameType == "unsubscribe" {
		if err := decodeStringField(object, "pattern", &frame.Pattern); err != nil {
			return ClientFrame{}, &Error{Code: "invalid_pattern", Message: "pattern must be a string"}
		}
		if err := ValidatePattern(frame.Pattern); err != nil {
			return ClientFrame{}, err
		}
		return frame, nil
	}
	if err := decodeStringField(object, "channel", &frame.Channel); err != nil {
		return ClientFrame{}, &Error{Code: "invalid_channel", Message: "channel must be a string"}
	}
	if err := ValidateChannel(frame.Channel); err != nil {
		return ClientFrame{}, err
	}
	value, ok := object["value"]
	if !ok {
		return ClientFrame{}, &Error{Code: "malformed_frame", Message: fmt.Sprintf("%s frame missing value", frameType)}
	}
	frame.Value = append(json.RawMessage(nil), value...)
	if traceRaw, ok := object["trace_id"]; ok {
		if err := json.Unmarshal(traceRaw, &frame.TraceID); err != nil || frame.TraceID == "" {
			return ClientFrame{}, &Error{Code: "malformed_frame", Message: "trace_id must be a non-empty string"}
		}
	}
	if depthRaw, ok := object["depth"]; ok {
		decoder := json.NewDecoder(bytes.NewReader(depthRaw))
		decoder.UseNumber()
		var value any
		if err := decoder.Decode(&value); err != nil {
			return ClientFrame{}, &Error{Code: "malformed_frame", Message: "depth must be a non-negative integer"}
		}
		number, ok := value.(json.Number)
		if !ok {
			return ClientFrame{}, &Error{Code: "malformed_frame", Message: "depth must be a non-negative integer"}
		}
		depth, err := number.Int64()
		if err != nil || depth < 0 || int64(int(depth)) != depth {
			return ClientFrame{}, &Error{Code: "malformed_frame", Message: "depth must be a non-negative integer"}
		}
		frame.Depth = int(depth)
	}
	return frame, nil
}

func rawObject(data []byte) (map[string]json.RawMessage, error) {
	var object map[string]json.RawMessage
	if err := json.Unmarshal(data, &object); err != nil || object == nil {
		return nil, errors.New("not an object")
	}
	return object, nil
}

func decodeStringField(object map[string]json.RawMessage, name string, target *string) error {
	raw, ok := object[name]
	if !ok {
		return errors.New("missing")
	}
	return json.Unmarshal(raw, target)
}
